src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs

```diff
@@ -1064,304 +907,250 @@ internal TransitionRegex<S> CreateDerivative()
             return _transitionRegex;
         }
 
-        // These are the cache for transition regexes with effects
-        private TransitionRegex<S>? _transitionRegexWithEffectsEager;
-        private TransitionRegex<S>? _transitionRegexWithEffectsAny;
-        private ref TransitionRegex<S>? TransitionRegexWithEffects(bool eager) => ref eager ? ref _transitionRegexWithEffectsEager : ref _transitionRegexWithEffectsAny;
+        /// <summary>
+        /// Takes the derivative of the symbolic regex for the given element, which must be either
+        /// a minterm (i.e. a class of characters that have identical behavior for all predicates in the pattern)
+        /// or a singleton set. This derivative is "eager", i.e. it only considers paths that backtracking would
+        /// take before accepting the empty string for this pattern and returns the pattern ordered in the order backtracking
+        /// would explore paths. For example the derivative of a*ab for a is a*ab|b, while for a*?ab it is b|a*?ab.
+        /// </summary>
+        /// <param name="elem">given element wrt which the derivative is taken</param>
+        /// <param name="context">immediately surrounding character context that affects nullability of anchors</param>
+        /// <returns>the derivative</returns>
+        internal SymbolicRegexNode<S> CreateEagerDerivative(S elem, uint context)
+        {
+            List<(SymbolicRegexNode<S>, DerivativeEffect[]?)> transitions = new();
+            AddTransitions(elem, context, transitions, new List<SymbolicRegexNode<S>>(), null);
+            SymbolicRegexNode<S> derivative = _builder._nothing;
+            for (int i = transitions.Count - 1; i >= 0; --i)
```

Comment:
Can you add a comment for why this iterates backwards and has the previously computed derivative on the rhs of the ordered or?